runtime/lua/vim/treesitter/dev.lua

```diff
@@ -436,11 +459,157 @@ function M.inspect_tree(opts)
     buffer = buf,
     once = true,
     callback = function()
-      if api.nvim_win_is_valid(w) then
-        api.nvim_win_close(w, true)
+      close_win(w)
+    end,
+  })
+end
+
+--- @class PreviewQueryOpts
+--- @field bufnr integer? Buffer to draw the editor into. If omitted, a new
+---                       buffer is created.
+--- @field winid integer? Window id to display the editor buffer in. If omitted,
+---                       a new window is created with {command}.
+--- @field command string? Vimscript command to create the window. Default
+---                        value is "60vnew" or "new" if the inspector is open.
+---                        Only used when {winid} is nil.
+
+--- @private
+--- @param opts PreviewQueryOpts?
+function M.preview_query(opts)
+  vim.validate({
+    opts = { opts, 't', true },
+  })
+
+  opts = opts or {}
+
+  local buf = api.nvim_get_current_buf()
+
+  -- Close any existing previewer window
+  if vim.b[buf].dev_preview then
+    close_win(vim.b[buf].dev_preview)
+  end
+
+  local w = opts.winid
+  if not w then
+    local cmd = opts.command or '60vnew'
+    -- If the inspector is open, place the previewer above it.
+    local base_buf = vim.b[buf].dev_base ---@type integer?
+    local inspect_buf = base_buf and vim.b[base_buf].dev_inspect
+    if base_buf and api.nvim_win_is_valid(inspect_buf) then
+      vim.api.nvim_set_current_win(inspect_buf)
+      buf = base_buf
+      cmd = 'new'
+    end
+    vim.cmd(cmd)
+    w = api.nvim_get_current_win()
+  end
+
+  local ok, parser = pcall(vim.treesitter.get_parser, buf)
+  if not ok then
+    return nil, 'No parser available for the given buffer'
+  end
+  local lang = parser:lang()
+
+  local b = opts.bufnr
+  if b then
+    api.nvim_win_set_buf(w, b)
+  else
+    b = api.nvim_win_get_buf(w)
+  end
+
+  vim.b[buf].dev_preview = w
+  vim.bo[b].omnifunc = 'v:lua.vim.treesitter.query.omnifunc'
+  set_dev_properties(w, b)
+
+  -- Note that omnifunc guesses the language based on the containing folder,
+  -- so we add the parser's language to the buffer's name so that omnifunc
+  -- can infer the language later.
+  api.nvim_buf_set_name(b, string.format('%s/query_previewer.scm', lang))
+
+  local ns = api.nvim_create_namespace('treesitter/dev-preview')
+
+  local function update_highlights()
+    api.nvim_buf_clear_namespace(buf, ns, 0, -1)
+    local query_content = table.concat(api.nvim_buf_get_lines(b, 0, -1, false), '\n')
+    local ok_query, query = pcall(vim.treesitter.query.parse, lang, query_content)
+    if ok_query then
+      local cursor_word = vim.fn.expand('<cword>')
+      -- Only highlight captures if the cursor is on a capture name
+      if
+        string.find(cursor_word --[[@as string]], '^@') == nil
+      then
+        return
+      end
+      -- Remove the '@' from the cursor word
+      cursor_word = string.sub(cursor_word --[[@as string]], 2)
+      for id, node in query:iter_captures(parser:trees()[1]:root(), buf, 0, -1) do
```

Comment:
This is setting highlights for the entire buffer which can be very slow.

Ideally we should use the decoration provider so we only highlight lines which are drawn.

Reply:
Thinking about this. We can probably just get away with limiting the query range here since we only place the highlights when the query editor is focused.

Reply:
I limited it to the visible range. Lmk what you think :)

Reply:
AHHH I'm silly. I restricted the wrong thing. I'll fix it.

Reply:
Okay I think I got it now.